Add support for Bazel

jwnimmer-tri opened the pull request:
Initial code review PR for https://github.com/lcm-proj/lcm/pull/560.

---

Add examples/bazel with a sample downstream use, licensed as MIT-0.

Other fixes:
- Adjust lcm_eventlog.py test to not require copying the example log.
- Fix a small typo in the go example.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jwnimmer-tri/lcm/1)
<!-- Reviewable:end -->

xuchen-han reviewed (commented):
Checkpoint after migrating over from upstream.

> Reviewed 25 of 51 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-:-OJA4wJP6QlIUHHdFgYU:b7gjli2)* status: 25 of 51 files reviewed, 3 unresolved discussions

___
*[`lcm-bazel/lcm_library.bzl` line 8 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4Cn9BQUNj0Du-Hrx:-OJA4Cn9BQUNj0Du-Hry:b32e8ca) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L8)):*
> ```Starlark
> lcm_library(
>     name = "foo_bar",
>     lcm_package = "foo.bar",
> ```

nit, from this example and the naming, it's not clear to me what the `lcm_package` field does. There's also an implicit requirement that the packages are delimited using ".". Consider making that explicit.
___
*[`lcm-bazel/lcm_library.bzl` line 23 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4_JgAy0Tb32eWqp6:-OJA4_JgAy0Tb32eWqp7:b9m41h8) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L23)):*
> ```Starlark
>     name = "foo_bar_c",
>     srcs = [":foo_bar_c_srcs"],
>     deps = ["@lcm//lcm:lcm-static"],
> ```

nit, It's not clear to me from just reading this example why this dep is needed.
___
*[`lcm-bazel/lcm_library.bzl` line 27 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4VgrD8l44XqR-89w:-OJA4VgrD8l44XqR-89x:b29oc7l) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L27)):*
> ```Starlark
> """
> 
> load("//lcm-bazel/private:lcm_library.bzl", _lcm_library = "lcm_library")
> ```

BTW, can you explain what this indirection achieves? Is it just for readabilty?


<!-- Sent from Reviewable.io -->

xuchen-han reviewed (commented):
Ok, almost done but I'm timed out for now. Will come back and finish soon.

> Reviewed 16 of 51 files at r1.
> *[Reviewable](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-:-OJAMuHSEVeeIlOstpMG:byj4qz9)* status: 41 of 51 files reviewed, 8 unresolved discussions (waiting on @jwnimmer-tri)

___
*[`lcm-bazel/lcm_library.bzl` line 23 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4_JgAy0Tb32eWqp6:-OJAAvpcACFuhDpDJhJ-:b-np374) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L23)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

nit, It's not clear to me from just reading this example why this dep is needed.
</blockquote></details>

Ok, this is more clear as I read more; but still would be nice to have some inline comments.
___
*[`lcm/BUILD.bazel` line 47 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA85aXD8GQwaAtxgUV:-OJA85aXD8GQwaAtxgUW:b90kbj5) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm/BUILD.bazel#L47)):*
> ```Starlark
> 
> LCM_LINKOPTS_WIN32 = [
>     # TODO(jwnimmer-tri) Not sure yet how to do this in Bazel.
> ```

BTW, would this affect building with bazel on WIN32 machines?
___
*[`lcm-bazel/private/lcm_c_library_srcs.bzl` line 10 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA73av5bRy7Yp9tlYR:-OJA73av5bRy7Yp9tlYS:b-pgisg8) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/private/lcm_c_library_srcs.bzl#L10)):*
> ```Starlark
>     # Declare the output files.
>     output_files = [
>         ctx.actions.declare_file("_".join(info.lcm_package_split + [x + ".c"]))
> ```

BTW are .c and .h files, unlike other generated files, not put into a subdir?
___
*[`.github/workflows/test_bazel.yml` line 1 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJAJFqgDhs_pjAO_7A_:-OJAJFqh0rD5vKbTzfbM:begu3h0) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/.github/workflows/test_bazel.yml#L1)):*
> ```YAML
> name: test_bazel
> ```

BTW, I won't be able to give a useful review on this. If you need an additional pair of eyes on this file, you should recruit another reviewer. If you think the tests passing is good enough evidence, I'm fine with that too.
___
*[`test/c/BUILD.bazel` line 123 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJALzzC8-V-iG8Q66gh:-OJALzzC8-V-iG8Q66gi:b-4128u1) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/test/c/BUILD.bazel#L123)):*
> ```Starlark
>         "LCM_DEFAULT_URL": "udpm://239.255.76.67:7601?ttl=0",
>     },
>     flaky = True,
> ```

BTW, I'm assuming this is following how these tests are handled in other build systems?
___
*[`lcm-python/BUILD.bazel` line 20 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJAIwllE0JacYX5ZDjO:-OJAIwllE0JacYX5ZDjP:blr22qk) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-python/BUILD.bazel#L20)):*
> ```Starlark
>     ],
>     linkshared = True,
>     linkstatic = True,
> ```

BTW, isn't this on by default for cc_binary?



_Code quote:_
```Starlark
linkstatic = True,
```


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/jwnimmer-tri/lcm/1)* status: 39 of 51 files reviewed, 5 unresolved discussions (waiting on @xuchenhan-tri)

___
*[`lcm/BUILD.bazel` line 47 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA85aXD8GQwaAtxgUV:-OJEPxP-B2g9ujxar46R:b-sf8eal) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm/BUILD.bazel#L47)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, would this affect building with bazel on WIN32 machines?
</blockquote></details>

Yes.  At the moment, it's probable that Windows doesn't build correctly with Bazel.
___
*[`lcm-bazel/lcm_library.bzl` line 8 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4Cn9BQUNj0Du-Hrx:-OJEScLA6ZKjdFSiDcIm:b-277h9n) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L8)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

nit, from this example and the naming, it's not clear to me what the `lcm_package` field does. There's also an implicit requirement that the packages are delimited using ".". Consider making that explicit.
</blockquote></details>

Added comments.
___
*[`lcm-bazel/lcm_library.bzl` line 23 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4_JgAy0Tb32eWqp6:-OJESe8c7q48Z1gs_hhk:b-277h9n) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L23)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Ok, this is more clear as I read more; but still would be nice to have some inline comments.
</blockquote></details>

Added comments.
___
*[`lcm-bazel/lcm_library.bzl` line 27 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4VgrD8l44XqR-89w:-OJEQ4k83-U6bhEJhRGW:btgorwd) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, can you explain what this indirection achieves? Is it just for readabilty?
</blockquote></details>

It's a convention by rules_foo authors to separate the public API from the private implementation in Skylark.  Anything that users should touch goes into bzl files in this directory.  Anything that is internal-use-only goes into the internal subdir.
___
*[`lcm-bazel/private/lcm_c_library_srcs.bzl` line 10 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA73av5bRy7Yp9tlYR:-OJEQIGDD9nZutyQnKQm:b-uvcctr) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/private/lcm_c_library_srcs.bzl#L10)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW are .c and .h files, unlike other generated files, not put into a subdir?
</blockquote></details>

Right!  That's how the upstream logic works.
___
*[`test/c/BUILD.bazel` line 123 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJALzzC8-V-iG8Q66gh:-OJEQtgl51df2Fw4H228:bxh2u4k) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/test/c/BUILD.bazel#L123)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, I'm assuming this is following how these tests are handled in other build systems?
</blockquote></details>

I think the other build systems just allow/expect the race conditions to happen and turn the test result red, and then developers just run the tests again by hand.


<!-- Sent from Reviewable.io -->

xuchen-han reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for the patience.

> Reviewed 9 of 51 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-:-OJFZtCMCchGHz-rebmy:b-dcb6c1)* status: 50 of 51 files reviewed, 2 unresolved discussions (waiting on @jwnimmer-tri)

___
*[`lcm/BUILD.bazel` line 47 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA85aXD8GQwaAtxgUV:-OJFZLtr3LfdGsxhz4mA:b-84xp0s) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm/BUILD.bazel#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes.  At the moment, it's probable that Windows doesn't build correctly with Bazel.
</blockquote></details>

Should we make a note of that in build_instructions.md?
___
*[`lcm-bazel/lcm_library.bzl` line 27 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA4VgrD8l44XqR-89w:-OJFZfMnBnv0z_eA2rLo:b-4hvi0m) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm-bazel/lcm_library.bzl#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It's a convention by rules_foo authors to separate the public API from the private implementation in Skylark.  Anything that users should touch goes into bzl files in this directory.  Anything that is internal-use-only goes into the internal subdir.
</blockquote></details>

Got it.
___
*[`test/c/BUILD.bazel` line 123 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJALzzC8-V-iG8Q66gh:-OJFZl8t5YwQStZXtHmH:bb7m7o0) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/test/c/BUILD.bazel#L123)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think the other build systems just allow/expect the race conditions to happen and turn the test result red, and then developers just run the tests again by hand.
</blockquote></details>

ouch


<!-- Sent from Reviewable.io -->

jwnimmer-tri committed:
Add support for Bazel

Add examples/bazel with a sample downstream use, licensed as MIT-0.

Other fixes:
- Adjust lcm_eventlog.py test to not require copying the example log.
- Fix a small typo in the go example.

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/jwnimmer-tri/lcm/1)* status: 50 of 51 files reviewed, all discussions resolved (waiting on @xuchenhan-tri)

___
*[`.github/workflows/test_bazel.yml` line 1 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJAJFqgDhs_pjAO_7A_:-OJUEq171WxvZUE-oZPD:b-vabl0i) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/.github/workflows/test_bazel.yml#L1)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, I won't be able to give a useful review on this. If you need an additional pair of eyes on this file, you should recruit another reviewer. If you think the tests passing is good enough evidence, I'm fine with that too.
</blockquote></details>

Thanks, my plan here is to ask the LCM upstream to review the CI part.
___
*[`lcm/BUILD.bazel` line 47 at r1](https://reviewable.io/reviews/jwnimmer-tri/lcm/1#-OJA85aXD8GQwaAtxgUV:-OJUFh578VB_hBMCEdl-:bb6un1j) ([raw file](https://github.com/jwnimmer-tri/lcm/blob/2daa671ee7b08a0f910330d2d415a9e6ff1c1dff/lcm/BUILD.bazel#L47)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Should we make a note of that in build_instructions.md?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->
